### PR TITLE
Add EP Support; Unified Parallelism Handling and Task-Based Architecture

### PR DESCRIPTION
**Pull request by yaoyu-33**
This PR introduces a major refactoring of the Megatron Bridge model conversion system, focusing on unifying parallelism handling (TP, PP, EP) and transitioning from a plan-based to a task-based architecture. The changes primarily affect weight conversion between HuggingFace and Megatron formats, with improved support for Mixture of Experts (MoE) models.

### 🏗️ **Architecture Refactoring**

- **Task-based conversion**: Replaced "plans" with "tasks" throughout the codebase
  - `_build_plan_hf_to_megatron()` → `_build_conversion_tasks()`
  - `_build_plan_megatron_to_hf()` → unified into `_build_conversion_tasks()`
  - Updated documentation and terminology to reflect task-based approach

- **Unified task handling**: Single `_build_conversion_tasks()` method now handles both HF→Megatron and Megatron→HF conversions
  - Improved parameter ordering based on HF source keys
  - Better handling of tied embeddings
  - Centralized task creation with consistent global parameter name resolution

#### **Pipeline Parallel (PP) Improvements**
- **Dynamic layer numbering**: Replaced static `get_transformer_layer_offset()` with dynamic `layer_module.layer_number` lookup
- **Robust parameter resolution**: New `_megatron_local_name_to_global()` function handles both PP and EP numbering
- **Global parameter collection**: Added `_megatron_global_param_names_all_pp_ranks()` with caching for efficient cross-rank parameter discovery

#### **Expert Parallel (EP) Integration**
- **Removed `MOEMapping` class**: Simplified expert handling by integrating EP logic directly into base mapping classes
- **Expert-aware parallelism**: All mapping classes now automatically detect and handle expert parameters via `is_expert` property
- **Expert gathering**: New `gather_from_ep_ranks()` method in `MegatronParamMapping` for collecting expert weights across EP ranks
- **Expert numbering**: Automatic conversion between local and global expert numbering in parameter names

#### **Tensor Parallel (TP) Enhancements**
- **Expert-aware TP groups**: Dynamic TP group selection based on parameter type (regular vs expert)
- **Improved parameter detection**: Better module and parameter resolution with `get_module_and_param_from_name()`
- **Enhanced bias handling**: Fixed row-parallel bias splitting for 1D tensors

### 🏷️ **Parameter Mapping Simplification**

- **Cleaner interfaces**: Removed `megatron_param_name` parameter from mapping methods
- **Unified parallelism properties**: Consistent `tp_rank`, `tp_size`, `ep_rank`, `ep_size`, etc. across all mappings
- **Process group utilities**: Leveraged `get_pg_rank()` and `get_pg_size()` for cleaner parallelism queries
- **Auto-detection improvements**: Enhanced `AutoMapping` to support `TopKRouter` and grouped linear layers

### 🔧 **Utility Functions**

- **Enhanced parameter resolution**: `get_module_and_param_from_name()` supports both absolute and relative parameter paths
- **Improved sorting**: `extract_sort_key()` for consistent parameter ordering based on layer and expert numbers
- **Removed legacy code**: Eliminated complex `get_transformer_layer_offset()` in favor of direct layer number access

### 📊 **Progress Tracking & UX**

- **Extracted progress wrapper**: `_with_progress_tracking()` method for consistent progress reporting
- **Simplified conversion loops**: Cleaner code structure with better error handling
- **Better tied embeddings handling**: Improved logic for models with shared embeddings and output weights

### **Expert Parallel Processing**
```python
# Before: Complex MOEMapping class with separate EP handling
class MOEMapping(MegatronParamMapping[torch.Tensor]): ...

# After: Integrated EP handling in base classes
@property
def is_expert(self) -> bool:
    return ".mlp.experts.linear_fc" in self.megatron_param

def gather_from_ep_ranks(self, megatron_weights, megatron_module, hf_param_name):
    # Unified expert gathering logic
```

### **Dynamic Layer Resolution**
```python
# Before: Static offset calculation
layer_offset = get_transformer_layer_offset(config, pipeline_rank, vp_stage)
global_layer_number = local_layer_number + layer_offset

# After: Dynamic module-based lookup
_, layer_module = get_module_and_param_from_name(models, layer_prefix, vp_stage)
global_layer_number = layer_module.layer_number - 1
```

### **Unified Task Creation**
```python
# Before: Separate plan builders for each direction
def _build_plan_hf_to_megatron(...)
def _build_plan_megatron_to_hf(...)

# After: Single task builder for both directions
def _build_conversion_tasks(self, hf_pretrained, megatron_model) -> List[WeightConversionTask]:
```

### CODE REVIEW

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 1]
```diff
@@ -103,68 +102,50 @@ class WeightConversionTask(Generic[MappingT]):
...
+                f".weight{local_expert_number}",
+                f".weight{global_expert_number}",
+            )
+        elif ".bias" in param_name:
+            local_expert_number = int(param_name.split(".bias")[-1])
+            global_expert_number = num_experts_per_rank * ep_group.rank() + local_expert_number
+            param_name = param_name.replace(
+                f".bias{local_expert_number}",
+                f".bias{global_expert_number}",
+            )
```
Comment (ananthsub) [inline src/megatron/bridge/models/conversion/model_bridge.py]: nit: code is duplicated except for the weight vs bias key. make it into a function to share the logic?

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 2]
```diff
@@ -346,53 +404,37 @@ def load_weights_hf_to_megatron(
...
-                            f"  Expected shape: {task.param_weight.shape}\n"
-                            f"  Got shape: {local_weights.shape}\n"
-                            f"  Bridge type: {type(task.mapping).__name__}\n"
-                            f"  HF mapping: {task.mapping.hf_param}"
-                        )
-                    task.param_weight.data.copy_(local_weights)
+        description = f"Loading from {hf_pretrained.model_name_or_path}"
+        for task in self._with_progress_tracking(hf_to_megatron_tasks, description):
+            if task.megatron_module is None:
+                continue
```
Comment (ananthsub) [inline src/megatron/bridge/models/conversion/model_bridge.py:416]: could you add a comment for when this could be None?

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 3]
```diff
@@ -414,39 +488,61 @@ def _get_config(self, module: nn.Module) -> Any:
...
+
+        # Extract local expert number from parameter name
+        # Handle both .weight and .bias suffixes
+        local_expert_number = None
+        if ".weight" in self.megatron_param:
+            global_expert_number = int(self.megatron_param.split(".weight")[-1])
+            local_expert_number = global_expert_number % num_experts_per_rank
+        elif ".bias" in self.megatron_param:
+            global_expert_number = int(self.megatron_param.split(".bias")[-1])
+            local_expert_number = global_expert_number % num_experts_per_rank
```
Comment (ananthsub) [inline src/megatron/bridge/models/conversion/param_mapping.py]:
```suggestion
        for key in (".weight", ".bias"):
            if key in self.megatron_param:
                global_expert_number = int(self.megatron_param.split(key)[-1])
                local_expert_number = global_expert_number % num_experts_per_rank
```

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 4]
```diff
@@ -524,13 +618,13 @@ def hf_to_megatron(
...
         hf_weights: torch.Tensor,
         megatron_module: nn.Module,
-        megatron_param_name: Optional[str] = None,
     ) -> torch.Tensor:
         """Split weight along dim 0 and distribute to TP ranks."""
         if self.tp_size == 1:
             return hf_weights
 
-        target_param, output_shape = _get_target_param_and_shape(self.megatron_param, megatron_module)
+        normalized_param = self._normalize_expert_param_name(self.megatron_param)
```
Comment (ananthsub) [inline src/megatron/bridge/models/conversion/param_mapping.py:636]: could you explain why we need to normalize for EP here?

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 5]
```diff
@@ -524,13 +625,16 @@ def hf_to_megatron(
...
-        megatron_param_name: Optional[str] = None,
     ) -> torch.Tensor:
         """Split weight along dim 0 and distribute to TP ranks."""
         if self.tp_size == 1:
             return hf_weights
 
-        target_param, output_shape = _get_target_param_and_shape(self.megatron_param, megatron_module)
+        # Some parameters are named with global expert number, e.g. experts.weight15,
+        # normalize it to experts.weight0, note we are only use the shape, dtype, device info,
+        # not the actual value, so it is safe to do this.
```
Comment (ananthsub) [inline src/megatron/bridge/models/conversion/param_mapping.py:635]:
```suggestion
        # Some parameters are named with global expert number, e.g. experts.weight15
        # normalize it to experts.weight0
        # Note we are only using the shape, dtype, device info,
        # not the actual value, so it is safe to do this.
```
